### PR TITLE
HPCC-19208 CSmartSocketFactory ctor parse out http[s]://

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

QA suite passes as expected

### CODE REVIEW

**Comment by mckellyln** [comment]
@rpastrana please review

File: system/jlib/jsmartsock.cpp [hunk 1]
```diff
@@ -45,9 +45,22 @@ class SmartSocketListParser
         unsigned port = defport;
 
         char *saveptr;
-        char *ip = strtok_r(copyFullText, "|", &saveptr);
-        while (ip != NULL)
+        char *ip0 = strtok_r(copyFullText, "|", &saveptr);
```
Comment (rpastrana) [inline system/jlib/jsmartsock.cpp]: preexisting, but I like to encourage explicit/accurate/selfdocumenting var names...

File: system/jlib/jsmartsock.cpp [hunk 2]
```diff
@@ -45,9 +45,22 @@ class SmartSocketListParser
...
+        while (ip0 != NULL)
```
Comment (rpastrana) [inline system/jlib/jsmartsock.cpp]: nullptr

File: system/jlib/jsmartsock.cpp [hunk 3]
```diff
@@ -45,9 +45,22 @@ class SmartSocketListParser
...
         {
+            // strip off http[s]://
+            char *ip = strstr(ip0, "http://");
```
Comment (rpastrana) [inline system/jlib/jsmartsock.cpp]: I can't remember the call, but I think there's a jlib function to split a connection string into prot;user;pass;host;port;uri, might be worth exploring

File: system/jlib/jsmartsock.cpp [hunk 4]
```diff
@@ -45,9 +45,22 @@ class SmartSocketListParser
...
+            if (ip != NULL)
```
Comment (rpastrana) [inline system/jlib/jsmartsock.cpp]: nullptr

**Review by rpastrana — COMMENTED** [review]
@mckellyln a couple of comments

File: system/jlib/jsmartsock.cpp (hunk 1 above)
Comment (mckellyln) [reply system/jlib/jsmartsock.cpp]: changed to hostportstr

File: system/jlib/jsmartsock.cpp (hunk 2 above)
Comment (mckellyln) [reply system/jlib/jsmartsock.cpp]: changed to nullptr

File: system/jlib/jsmartsock.cpp (hunk 4 above)
Comment (mckellyln) [reply system/jlib/jsmartsock.cpp]: changed to nullptr

File: system/jlib/jsmartsock.cpp (hunk 3 above)
Comment (mckellyln) [reply system/jlib/jsmartsock.cpp]: I don't know of a routine to handle the ip range like this does (ie 1.2.3.4-6:8000).

**Comment by mckellyln** [comment]
@rpastrana changed var name to be more descriptive and NULL to nullptr in new code.
I could not find a suitable routine to easily put in that can parse the ip range so I left it as is.

**Review by rpastrana — APPROVED** [review]
@mckellyln

**Comment by mckellyln** [comment]
@rpastrana added check for HTTP[S}:// as well.

**Review by rpastrana — APPROVED** [review]
@mckellyln ok to merge.

**Pull request merged**